File: python/aoc2022/days/day8.py
def day8():
    with open('../input/input8.txt', 'r') as f:
        data = f.read()

    forest_map = []
    visible_map = []

    for index_r, row in enumerate(data.splitlines()):
        maprow = []
        visiblerow = []
        for index_c, col in enumerate(row):
            maprow.append(int(col))
            if index_c == 0 or index_r == 0:
                visiblerow.append(1)
            else:
                visiblerow.append(0)
        forest_map.append(maprow)
        visible_map.append(visiblerow)

    width = len(forest_map[0])
    height = len(forest_map)

    for i in range(1, height - 1):
        curmaxl = 0
        curmaxr = 0
        for j in range(1, width - 1):
            if forest_map[i][j - 1] > curmaxl:
                curmaxl = forest_map[i][j - 1]
            if curmaxl < forest_map[i][j]:
                visible_map[i][j] = 1
            if forest_map[(height - 1) - i][(width - 1) - (j - 1)] > curmaxr:
                curmaxr = forest_map[(height - 1) - i][(width - 1) - (j - 1)]
            if curmaxr < forest_map[(height - 1) - i][(width - 1) - j]:
                visible_map[(height - 1) - i][(width - 1) - j] = 1

    for j in range(1, width - 1):
        curmaxl = 0
        curmaxr = 0
        for i in range(1, height - 1):
            if forest_map[i - 1][j] > curmaxl:
                curmaxl = forest_map[i - 1][j]
            if curmaxl < forest_map[i][j]:
                visible_map[i][j] = 1
            if forest_map[(height - 1) - (i - 1)][(width - 1) - j] > curmaxr:
                curmaxr = forest_map[(height - 1) - (i - 1)][(width - 1) - j]
            if curmaxr < forest_map[(height - 1) - i][(width - 1) - j]:
                visible_map[(height - 1) - i][(width - 1) - j] = 1

    best = 0
    for i in range(height):
        for j in range(width):
            score = view(i, j, forest_map)
            if score > best:
                best = score

    print('Part 1: ', sum([sum([col for col in row]) for row in visible_map]))
    print('Part 2: ', best)


def view(i, j, forest_map):
    scorerl = 0
    scorerr = 0
    scorecl = 0
    scorecr = 0
    for r in range(1, i + 1):
        if forest_map[i - r][j] < forest_map[i][j]:
            scorerl += 1
        else:
            scorerl += 1
            break
    for r in range(1, len(forest_map) - i):
        if forest_map[i + r][j] < forest_map[i][j]:
            scorerr += 1
        else:
            scorerr += 1
            break
    for c in range(1, j + 1):
        if forest_map[i][j - c] < forest_map[i][j]:
            scorecl += 1
        else:
            scorecl += 1
            break
    for c in range(1, len(forest_map) - j):
        if forest_map[i][j + c] < forest_map[i][j]:
            scorecr += 1
        else:
            scorecr += 1
            break

    return scorerl * scorerr * scorecr * scorecl

File: rust/src/aoc2022/day8.rs
use rdcl_aoc_helpers::input::WithReadLines;
use std::{fs::File};

pub fn day8(path: &String) {

    let data:Vec<String> = File::open(path).read_lines::<String>(1).collect();

    let treemap: Vec<Vec<u32>> = data.iter().map(|row| row.chars().map(|c| c.to_digit(10).unwrap()).collect()).collect();

    let height = treemap.len();
    let width = treemap[0].len();

    let mut visiblemap = treemap.clone();
    visiblemap.iter_mut().for_each(|row| row.iter_mut().for_each(|c| *c = 0));

    for i in 0..height {
        for j in 0..width {
            if i == 0 || j == 0 || i == height-1 || j == width-1 {
                visiblemap[i][j] = 1;
            }
        }
    }

    for i in 1..height-1 {
        let mut curmaxl:u32 = 0;
        let mut curmaxr:u32 = 0;
        for j in 1..width-1 {
            if treemap[i][j - 1] > curmaxl{
                curmaxl = treemap[i][j - 1];
            }
            if curmaxl < treemap[i][j]{
                visiblemap[i][j] = 1;
            }
            if treemap[(height - 1) - i][(width - 1) - (j - 1)] > curmaxr {
                curmaxr = treemap[(height - 1) - i][(width - 1) - (j - 1)];
            }
                
            if curmaxr < treemap[(height - 1) - i][(width - 1) - j]{
                visiblemap[(height - 1) - i][(width - 1) - j] = 1;
            }
        }
    }

    for j in 1..width-1 {
        let mut curmaxl:u32 = 0;
        let mut curmaxr:u32 = 0;
        for i in 1..height-1 {
            if treemap[i - 1][j] > curmaxl{
                curmaxl = treemap[i - 1][j];
            }
            if curmaxl < treemap[i][j]{
                visiblemap[i][j] = 1;
            }
            if treemap[(height - 1) - (i - 1)][(width - 1) - j] > curmaxr {
                curmaxr = treemap[(height - 1) - (i - 1)][(width - 1) - j];
            }
                
            if curmaxr < treemap[(height - 1) - i][(width - 1) - j]{
                visiblemap[(height - 1) - i][(width - 1) - j] = 1;
            }
        }
    }

    let mut best:u32 = 0;
    let mut score:u32;
    for i in 1..height-1{
        for j in 1..width-1{
            score = view(i, j, &treemap);
            if score > best{
                best = score;
            }
        }
    }

    println!("Part 1: {}", visiblemap.iter().map(|x| -> u32 { x.iter().sum() }).sum::<u32>());
    println!("Part 2: {}", best);
}

fn view(i:usize, j:usize, treemap:&Vec<Vec<u32>>) -> u32 {
    let mut scorerl:u32 = 0;
    let mut scorerr:u32 = 0;
    let mut scorecl:u32 = 0;
    let mut scorecr:u32 = 0;
    for r in 1..(i+1) {
        if treemap[i - r][j] < treemap[i][j]{
            scorerl += 1;
        } else {
            scorerl += 1;
            break
        }
    }
    for r in 1..(treemap.len() - i){
        if treemap[i + r][j] < treemap[i][j]{
            scorerr += 1;
        }
        else{
            scorerr += 1;
            break
        }
    }
    for c in 1..(j+1) {
        if treemap[i][j - c] < treemap[i][j] {
            scorecl += 1;
        } else {
            scorecl += 1;
            break;
        }
    }
    for c in 1..(treemap.len() - j) {
        if treemap[i][j + c] < treemap[i][j] {
            scorecr += 1;
        } else {
            scorecr += 1;
            break;
        }
    }


    let score = scorecl * scorecr * scorerl * scorerr;
    score
}